ginkgo/outline/ginkgo.go

```diff
@@ -0,0 +1,209 @@
+package outline
+
+import (
+	"go/ast"
+	"go/token"
+	"strconv"
+)
+
+const (
+	// undefinedTextAlt is used if the spec/container text cannot be derived
+	undefinedTextAlt = "undefined"
+)
+
+// ginkgoMetadata holds useful bits of information for every entry in the outline
+type ginkgoMetadata struct {
+	// Name is the spec or container function name, e.g. `Describe` or `It`
+	Name string `json:"name"`
+
+	// Text is the `text` argument passed to specs, and some containers
+	Text string `json:"text"`
+
+	// Start is the position of first character of the spec or container block
+	Start token.Pos `json:"start"`
+
+	// End is the position of first character immediately after the spec or container block
+	End token.Pos `json:"end"`
+
+	Spec    bool `json:"spec"`
+	Focused bool `json:"focused"`
+	Pending bool `json:"pending"`
+}
+
+// ginkgoNode is used to construct the outline as a tree
+type ginkgoNode struct {
+	ginkgoMetadata
+	Nodes []*ginkgoNode `json:"nodes,omitempty"`
+}
+
+type walkFunc func(n *ginkgoNode)
+
+func (n *ginkgoNode) PreOrder(f walkFunc) {
+	f(n)
+	for _, m := range n.Nodes {
+		m.PreOrder(f)
+	}
+}
+
+func (n *ginkgoNode) PostOrder(f walkFunc) {
+	for _, m := range n.Nodes {
+		m.PostOrder(f)
+	}
+	f(n)
+}
+
+func (n *ginkgoNode) Walk(pre, post walkFunc) {
+	pre(n)
+	for _, m := range n.Nodes {
+		m.Walk(pre, post)
+	}
+	post(n)
+}
+
+// PropagateInheritedProperties propagates the Pending and Focused properties
+// through the subtree rooted at n.
+func (n *ginkgoNode) PropagateInheritedProperties() {
+	n.PreOrder(func(thisNode *ginkgoNode) {
+		for _, descendantNode := range thisNode.Nodes {
+			if thisNode.Pending {
+				descendantNode.Pending = true
+				descendantNode.Focused = false
+			}
+			if thisNode.Focused && !descendantNode.Pending {
+				descendantNode.Focused = true
+			}
+		}
+	})
+}
+
+// BackpropagateUnfocus propagates the Focused property through the subtree
+// rooted at n. It applies the rule described in the Ginkgo docs:
+// > Nested programmatically focused specs follow a simple rule: if a
+// > leaf-node is marked focused, any of its ancestor nodes that are marked
+// > focus will be unfocused.
+func (n *ginkgoNode) BackpropagateUnfocus() {
+	focusedSpecInSubtreeStack := []bool{}
```

Comment:
This is hard to understand. I need to add some explanatory comments, and probably a unit test.

Reply:
the whole back propagate thing is intrinsically complex!  thanks for making this work!!